Deprecate generic stride implementation (#17812), allow AbstractArrays to use BLAS/LAPACK routines (#25247)

dlfivefifty opened the pull request:
This pull request resolves Issues #17812 and #25247:

1. Deprecate `stride(::AbstractArray, i::Int)` and `strides(::AbstractArray)`.
2. Re-implement above for the special cases of `StridedArray`.
3. Removes restrictions to `StridedArray` in BLAS routines, so that user types that are memory-equivalent to strided arrays can access these routines.
4. Removes the restrictions to `StridedArray` in `sparse/sparsevector.jl`. I'm not sure why these were there to begin.

I still need to weaken the types for ther LAPack routines. This is straightforward but will take some time to get the tests implemented.

Commit:
deprecate stride, support AbstractArray in blas routines

dlfivefifty committed:
move size_to_strides back to abstractarray.jl

dlfivefifty committed:
use parents stride in substrides

dlfivefifty committed:
Add strides for StridedReshapedArray

dlfivefifty committed:
Move stride, stride1 to dense.jl

dlfivefifty committed:
move strides tests to test/dense.jl

stevengj commented:
So you need a subtype of `StridedArray` to use BLAS?   Wouldn't it be better to pull away from subtypes here and push towards traits?

dlfivefifty commented:
That’s exactly what this is doing: the strides matrix interface consists of overriding conversion to ` Ptr ` and implementing ` stride `.

Down the line a trait could be added as well but this is the first step.

stevengj commented on base/abstractarray.jl:
```diff
@@ -258,7 +249,20 @@ julia> strides(A)
-strides(A::AbstractArray) = size_to_strides(1, size(A)...)
+function strides end
+
+# the definition of strides for Array is the cumsum of the product of sizes
```
cumprod of the size?  But it's not quite that, it's `cumprod([1;collect(size(A))[1:end-1]])`

stevengj commented on doc/src/manual/interfaces.md:
```diff
@@ -391,6 +391,24 @@ something other than 1), you should specialize `indices`. You should also specia
+Provided the element type of the array is compatible with BLAS, a strided array can utilize BLAS and LAPACK routines
+for more efficient linear algebra routines.
+
+A typical example of a user defined strided array is one that wraps a standard `Array`
```
hyphenate "user-defined" (hyphenate compound adjectives).  Don't hyphenate "subtype".

dlfivefifty committed:
_cumsumprodsizes -> _defaultstride, addresses stevengl comments

stevengj commented on base/abstractarray.jl:
```diff
@@ -251,8 +251,8 @@ julia> strides(A)
 
-# the definition of strides for Array is the cumsum of the product of sizes
-function _cumsumprodsizes(a::AbstractArray, i)
+# the definition of strides for Array is cumprod([1;collect(size(A))[1:end-1]])
```
well, technically `cumprod` would return an array, whereas strides returns a tuple

dlfivefifty replied:
Hmm, that's not quite right since:
```julia
julia> A = Array{Float64,0}()
0-dimensional Array{Float64,0}:
0.0

julia> strides(A)
()
```
How about saying:

> the defition of strides for `Array{T,N}` is `tuple()` if `N = 0`, otherwise it is a tuple containing 1 and a cumulative product of the first N-1 sizes

dlfivefifty committed:
Remove _defaultstride since it's only used one place

timholy commented on doc/src/manual/interfaces.md:
```diff
@@ -391,6 +391,24 @@ something other than 1), you should specialize `indices`. You should also specia
+A typical example of a user-defined strided array is one that wraps a standard `Array`
+with additional structure.
+
+
```
I'd add a warning here too about how dangerous it is to implement these methods if the underlying storage is *not* actually strided. Otherwise the following might happen:
- user calls a function. Gets a MethodError due to missing implementation of `strides`
- "helpfully" implements `strides` to fix the method error

It might be useful to give some concrete examples of arrays and categorize them as to whether they have strided memory representation. Examples:
```julia
1:5   # not strided (there is no storage associated with this array, could segfault if implemented `strides`)
collect(1:5)  # is strided
A = [1 5; 2 6; 3 7; 4 8]  # is strided
V = view(A, 1:2, :)   # is strided
V = view(A, 1:2:4, :)   # is strided
V = view(A, [1,2,4], :)   # is not strided
```

timholy reviewed (commented):
This is looking good, thanks for tackling this.

dlfivefifty commented on base/linalg/linalg.jl:
```diff
@@ -216,9 +216,40 @@ end
 # Check that stride of matrix/vector is 1
 # Writing like this to avoid splatting penalty when called with multiple arguments,
 # see PR 16416
+"""
```
This happens automatically at compile time? Cool!

I'm not sure how to test for performance regression or that it's doing the right behaviour, so I might hold off on that change (since it's independent of this PR).

dlfivefifty commented on base/subarray.jl:
```diff
@@ -255,11 +255,11 @@ IndexStyle(::Type{<:SubArray}) = IndexCartesian()
 strides(V::SubArray) = substrides(V.parent, V.indices)
 
-substrides(parent, I::Tuple) = substrides(1, parent, 1, I)
+substrides(parent, I::Tuple) = substrides(stride(parent, 1), parent, 1, I)
```
Good point. Matbe something like:
```julia
substrides(parent, I) = substrides(parent, strides(parent), I)
substrides(parent, strds, ::Tuple{}) = ()
substrides(parent, strds, I::Tuple{ScalarIndex, Vararg{Any}}) = (substrides(parent, tail(strds), tail(I))...,)
substrides(parent, strds, I::Tuple{Slice, Vararg{Any}}) = (first(strds), substrides(parent, tail(strds), tail(I))...)
substrides(parent, strds, I::Tuple{AbstractRange, Vararg{Any}}) = (first(strds)*step(I[1]), substrides(parent, tail(strds), tail(I))...)
substrides(parent, strds, I::Tuple{Any, Vararg{Any}}) = throw(ArgumentError("strides is invalid for SubArrays with indices of type $(typeof(I[1]))"))
```
Does `substrides` have to go through a deprecation process? It's not exported.

dlfivefifty committed:
Remove restriction to StridedArray in LAPACK

dlfivefifty committed:
Make strides(A) the default override, make sterner warning in deprecation message and docs

dlfivefifty commented:
I've changed this a bit so that the default override is `strides(A)` and so one does not need to override `stride(A,i)`. This is consistent with the abstract array interface, where one overrides `size(A)`, not `size(A,i)`.

I've also removed the restriction to `StridedArray` in the LAPACK routines, in the process I found and fixed some missing `chkstride1` calls, and a bug in `tzrzf!` which used `size` instead of `stride`.

dlfivefifty committed:
O(N) substrides implementation

dlfivefifty committed:
Deprecate old substrides

dlfivefifty commented:
There is an issue with type-inferrence with the new `substrides` implementation:
```julia
julia> versioninfo()
Julia Version 0.7.0-DEV.3204
Commit d1086e0091* (2017-12-30 21:09 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin17.3.0)
  CPU: Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, skylake)
Environment:
  JULIA_VERSION = 0.6

julia> A = rand(5,5,5,5);

julia> V = view(A, 1:1 ,:, 1:3, :);

julia> using Test

julia> @inferred strides(A)
(1, 5, 25, 125)

julia> @inferred strides(V)
ERROR: return type NTuple{4,Int64} does not match inferred return type Tuple{Int64,Int64,Int64,Any}
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] top-level scope
```
This seems to be a limitation on type inference and I'm not sure how to resolve it.


The relevant function is 
```julia
substrides(parent, I::Tuple) = substrides(parent, strides(parent), I)
substrides(parent, strds, ::Tuple{}) = ()
substrides(parent, strds, I::Tuple{ScalarIndex, Vararg{Any}}) = (substrides(parent, tail(strds), tail(I))...,)
substrides(parent, strds, I::Tuple{Slice, Vararg{Any}}) = (first(strds), substrides(parent, tail(strds), tail(I))...)
substrides(parent, strds, I::Tuple{AbstractRange, Vararg{Any}}) = (first(strds)*step(I[1]), substrides(parent, tail(strds), tail(I))...)
substrides(parent, strds, I::Tuple{Any, Vararg{Any}}) = throw(ArgumentError("strides is invalid for SubArrays with indices of type $(typeof(I[1]))"))
```

dlfivefifty committed:
work around type inference failure in substrides

dlfivefifty commented:
I've worked around the issue with `substrides` by adding an `::Int` assertion.

I think this PR is ready to be merged, though I'm happy to make any other changes. 

There is one test failure but it appears to be a download failure, which shouldn't be caused by this PR. It doesn't look like I have the ability to restart the test.

stevengj commented:
An `::Int` assertion may just be hiding the underlying problem; I'm worried that an `Any` variable is still heap-allocated somewhere and then converted to `Int`?

dlfivefifty commented:
Good point. I'm not sure how to proceed as it appears to be a breakdown of type inference. 

Perhaps putting the type assertion in the arguments is better? That is, for example:
```julia
substrides(parent, strds::NTuple{N,Int}, I::Tuple{Slice, Vararg{Any}}) where N = (first(strds), substrides(parent, tail(strds), tail(I))...)
```

timholy commented:
Yeah, this does seem to be an inference bug. I get different answers depending on sequence of operations. For example:
```julia
@code_warntype substrides(A, (1, 5, 25, 125), (1:1, Base.Slice(Base.OneTo(5)), 1:3, Base.Slice(Base.OneTo(5))))  # "fails"
@code_warntype substrides(A, (5, 25, 125), (Base.Slice(Base.OneTo(5)), 1:3, Base.Slice(Base.OneTo(5))))  # "succeeds"
@code_warntype substrides(A, (1, 5, 25, 125), (1:1, Base.Slice(Base.OneTo(5)), 1:3, Base.Slice(Base.OneTo(5)))) # "succeeds"
```

```

dlfivefifty committed:
state argument type in substrides so inference is correct

timholy commented on test/linalg/blas.jl:
```diff
@@ -326,3 +326,108 @@ end
+Base.setindex!(A::WrappedArray, v, i::Int) = setindex!(A.A, v, i)
+Base.setindex!(A::WrappedArray{T, N}, v, I::Vararg{Int, N}) where {T, N} = setindex!(A.A, v, I...)
+Base.unsafe_convert(::Type{Ptr{T}}, A::WrappedArray{T}) where T = Base.unsafe_convert(Ptr{T}, A.A)
+Base.stride(A::WrappedArray, i::Int) = stride(A.A, i)
```
Perhaps before this line you can test that `strides(A::WrappedArray)` and `stride(A::WrappedArray, i)` both throw an error, and that afterward it does not? In a sense, the main contribution of this PR is *removing* a fallback method, so we should test that the fallback doesn't accidentally get reintroduced.

Or you could define `WrappedArray2` and test for the error.

dlfivefifty replied:
I assume by "throws error" you mean `@test_deprecated` since the fallback is still there?

timholy replied:
Right indeed. :+1:

timholy reviewed (approved):
Presuming tests pass, one tiny tweak and I think this is done. Many, many thanks for seeing this through!

dlfivefifty committed:
merge master

dlfivefifty committed:
test_deprecated for strides before implementing

dlfivefifty commented:
I think it’s ready to be merged now (again). The only failed test was a download error, which I don’t think this PR could cause.

mbauman commented:
This is really nice! :+1:

timholy commented:
I belatedly realized that I should have asked you to add the deprecation to NEWS.md. Any chance you could do that now?

dlfivefifty commented:
Yes I’ll try to do this soon.

Nosferican commented:
The deprecation method for this ain't very helpful on 0.7, given this message what should the appropriate method to update code be?
```
 Warning: The default `strides(a::AbstractArray)` implementation is deprecated for general arrays.
│ Specialize `strides(::Adjoint)` if `Adjoint` indeed uses a strided representation in memory.
│ Warning: inappropriately implementing this method for an array type that does not use strided
│ storage may lead to incorrect results or segfaults.
│   caller = stride at abstractarray.jl:350 [inlined]
└ @ Core ./abstractarray.jl:350
```

timholy commented:
The alternative is to use cartesian indexing, i.e., `A[i,j,k]` rather than `A[i]` for a 3d array.

dlfivefifty commented:
If your custom array is strided in memory, then it should implement the interface and override `strides(::MyArray)`. 

If it's not strided in memory, you probably actually mean to use `size(A,1)`, not `stride(A,2)`.

timholy commented:
The xref above (added after my response) is from PyPlot, which probably means it *is* a pointer operation. Tread carefully in terms of following the advice above; I see there's an `Adjoint` in the message which means that the fallback implementation of `cumprod([1, size(A)...])[1:end-1]` may give the wrong answer.

The best advice might be to read up on strided arrays; once you understand the concept you will know if your array is strided and how to best implement the fix.

mbauman commented:
It may be that #29135 will solve the issue.  A second thumbs up there would be good and then it can be merged.
